### ISSUE TITLE
Adaptive: memory oom issue 

### ISSUE THREAD

**Issue by linking12** [netty/netty#14598]
This is a continuation of the discussion on Adpitve. The original thread has been closed.  https://github.com/netty/netty/issues/14553

I found that the rss memory will continue to increase. Although it will decrease after a certain period of time, it is generally increasing. Now I understand why the mmap memory why increase. It should be that when the rss is not enough, it will try to exchange memory through mmap
1: rss increase
<img width="1384" alt="image" src="https://github.com/user-attachments/assets/de4923bb-227c-474f-a401-07b6a291fba4" />
2: total memory increase
<img width="1071" alt="image" src="https://github.com/user-attachments/assets/a5834aec-85f9-4e43-828e-43533b10c035" />

i'm not sure this memory leak issue is fixed;

**Comment by franz1981** [comment]
I don't understand this...please try profiling the application via async profiler -e mprotect to see if there's anything on application side which keep on extending the existing malloc arenas.
Another option is to get the JDK system map at intervals via `jcmd System.map` , see https://bugs.openjdk.org/browse/JDK-8318636

and monitor which parts of the mapping change over time.

Another option, is to enable malloc autotrimming at intervals (if your hotspot version support it) and see if it helps.

**Comment by linking12** [comment]
> I don't understand this...please try profiling the application via async profiler -e mprotect to see if there's anything on application side which keep on extending the existing malloc arenas. Another option is to get the JDK system map at intervals via `jcmd System.map` , see https://bugs.openjdk.org/browse/JDK-8318636
> 
> and monitor which parts of the mapping change over time.
> 
> Another option, is to enable malloc autotrimming at intervals (if your hotspot version support it) and see if it helps.


<img width="1379" alt="image" src="https://github.com/user-attachments/assets/601b4ad2-bac9-429f-a3ac-1e3ef51edbd8" />

It is safe, just like @normanmaurer explained in the PR, the memory will be released in the end, but the monitoring looks scary, and I think it is necessary to dynamically scaling the RSS. @franz1981

**Comment by ww8912188** [comment]
Our service eventually went OOM after running for 16 hours. It tried to reclaim memory in the middle, but the overall trend continued to increase. Can I ask if we could have a more aggressive memory recycling strategy to avoid similar OOM.

In addition, we found another node also had a similar phenomenon of out-of-heap memory increase. Our JDK version did not support Jcmd System.map, so pmap was used. Please refer to the attached file.

<img width="1633" alt="image" src="https://github.com/user-attachments/assets/f6ccc1b4-3048-4702-bea7-6367ea8aa6e8" />

<img width="1655" alt="image" src="https://github.com/user-attachments/assets/4e70b3f1-f8e3-4eef-bb7c-b44848aca7d3" />

[pmap.txt](https://github.com/user-attachments/files/18192636/pmap.txt)

**Comment by linking12** [comment]
> Our service eventually went OOM after running for 16 hours. It tried to reclaim memory in the middle, but the overall trend continued to increase. Can I ask if we could have a more aggressive memory recycling strategy to avoid similar OOM.
> 
> In addition, we found another node also had a similar phenomenon of out-of-heap memory increase. Our JDK version did not support Jcmd System.map, so pmap was used. Please refer to the attached file.
> 
> <img alt="image" width="1633" src="https://private-user-images.githubusercontent.com/17057416/397225711-f6ccc1b4-3048-4702-bea7-6367ea8aa6e8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ1ODQyMjIsIm5iZiI6MTczNDU4MzkyMiwicGF0aCI6Ii8xNzA1NzQxNi8zOTcyMjU3MTEtZjZjY2MxYjQtMzA0OC00NzAyLWJlYTctNjM2N2VhOGFhNmU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE5VDA0NTIwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIzMWYxYzA1YWQ0MThmYzUwMjFiMjE5YjcwZjM3Zjg5NzVkMzVlMGM4MTEwZTQ4MWQ0ZTczZWJjZmJlN2U2YzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.8wxkeJ6YD8vfqwcJQpF98bJt4U0G2Bh6h4JEGisuHQ8"> <img alt="image" width="1655" src="https://private-user-images.githubusercontent.com/17057416/397226185-4e70b3f1-f8e3-4eef-bb7c-b44848aca7d3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ1ODQyMjIsIm5iZiI6MTczNDU4MzkyMiwicGF0aCI6Ii8xNzA1NzQxNi8zOTcyMjYxODUtNGU3MGIzZjEtZjhlMy00ZWVmLWJiN2MtYjQ0ODQ4YWNhN2QzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE5VDA0NTIwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY2MzBkZmEwODM2NjgwZjg1MTZmYTg4MzUxZDA4MWQ2OWM0MWVjZDlhZmZlMzI4YzMxMjI2ODhiOWMwNTE5ZDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.qdF05zVtEiJC2IIltgkRIZ-iLComqpxmvvYcvEuSpeQ">
> [pmap.txt](https://github.com/user-attachments/files/18192636/pmap.txt)

@normanmaurer @franz1981 

1: Is it too conservative to clean up the memory as time goes by? Is there a scheduled task to actively clean up? For example, when maxDirectbuffer is 80%, it is cleaned up. In fact, our application does not set maxDirectbuffer. Does it mean that Adaptive thinks there is enough space to apply for it, but does not actively release it?
2: Other information When we roll back to version 114, this problem will not have such a big impact, although the memory has increased;
(The application itself uses version 116 code, but there is another agent using version 114)

Looking forward to your reply；

**Comment by He-Pin** [comment]
I knew the Adaptive one was for performance, so despite the memory issue, do you see performance improvement, @linking12 @ww8912188?

**Comment by He-Pin** [comment]
>Is there a scheduled task to actively clean up? 
There is an issue open for that, you could say, on the roadmap of @franz1981

**Comment by linking12** [comment]
> I knew the Adaptive one was for performance, so despite the memory issue, do you see performance improvement, @linking12 @ww8912188?

This is what we expect. Our test shows that the performance throughput has increased by 5%, so we don’t want to give it up and go back to using pooling.

**Comment by He-Pin** [comment]
Thanks for sharing the number.

**Comment by franz1981** [comment]
@linking12 any news or new info about it?

**Comment by linking12** [comment]
> [@linking12](https://github.com/linking12) any news or new info about it?
> 
> Let me give you some insight after reading the pmap file which contains several anon mappings with large reserved areas (but not committed) - and that is usually a sign of this problem described at https://bugs.openjdk.org/browse/JDK-8193521
> 
> Unless these arenas are written, they won't commit any memory but still they can create tons of pressure indirectly because:
> 
> * if accessed in a sparse way can lead to fragmentation of the virtual address space
> * the OS page table grow larger and that consume RSS
> * the glibc malloc metadata to handle such grows larger, and that can contribute to consume RSS
> 
> As explained in the mentioned issue, this is likely happening because of a known inability of the OS glibc allocator of being container-friendly and indeed is a good choice to switch to jemalloc (which instead better handle these scenarios) or reduce the max areanas. It can lead to some contention, but if sized to match the actual dedicated cpu quotas will be much cheaper.
> 
> Last but not least, how this is happening - why it wasn't that bad before, and I have my educated guesses here as well:
> 
> 1. the jemalloc-based Netty (old) allocator usually allocate in bigger arenas - but if users uses worker thread pools (which can grow by injecting new threads) - it will still pick it (if there's some available) memory from existing arenas
> 2. adaptive instead could fall off the available buffers, because of contention - and eventually cause new thread local arenas to be reserved (with very few memory committed)
> 
> I have yet to figure it out exactly what could make this happen, really -and I'm assuming you have tons of allocations from some worker thread pools instead from event loops, but right now this is the best guess I have with this limited information.
> 
> Just switching to a different OS allocator or tuning the arenas max count will reveal if I'm right :)
> 
> NOTE: if, instead, this is just a leak we missed because of some programmatic error, I suggest to perform some mem leak analysis using jemalloc profiling (you have to switch to jemalloc) or use [async-profiler/async-profiler#1064](https://github.com/async-profiler/async-profiler/pull/1064) (memleak) or https://github.com/async-profiler/async-profiler/blob/d1498a6c7fda7c5987caf5e301c3de1deb9743c5/docs/ProfilingModes.md#native-memory-leaks . As the links shows, the new version of this leak profiler in async profiler doesn't requires LD_PRELOAD since it patchs the usage of malloc @https://github.com/async-profiler/async-profiler/blob/d1498a6c7fda7c5987caf5e301c3de1deb9743c5/src/mallocTracer.cpp#L99

@franz1981  thank you for your research, there is no news to report, We are taking the following step：
1: To avoid problems in the production environment, we switched adpitve to pooled
2: I also think the problem is caused by memory fragmentation, and we are going to switch glibc to jemalloc, which is currently in progress

**Comment by chrisvest** [comment]
@linking12 Did the problem persist despite switching from adaptive to pooled? Or am I reading too much into your message?

We recently also merged this, which might help the adaptive allocator: https://github.com/netty/netty/pull/14768

**Comment by linking12** [comment]
> [@linking12](https://github.com/linking12) Did the problem persist despite switching from adaptive to pooled? Or am I reading too much into your message?
> 
> We recently also merged this, which might help the adaptive allocator: [#14768](https://github.com/netty/netty/pull/14768)

Sorry for the late reply.  only adaptive have issue, pooled do not have issue;
I have noticed this PR, we are doing some testing, it will take some time

**Comment by franz1981** [comment]
It is now fine @linking12 ?? Awesome!

**Comment by He-Pin** [comment]
I'm preparing a new docker image, I will run stress teasting during this weekend.

**Comment by He-Pin** [comment]
<img width="1181" alt="Image" src="https://github.com/user-attachments/assets/a5a351c2-41f5-46f7-9225-15af8be9ae52" />
I tested the new allocator with two rounds of 10-hour stress. It seems the old pooled bytebuf allocator is more suitable for my SSR scene.

**Comment by franz1981** [comment]
@He-Pin I don't understand the .meaning of all the numbers there (e.g. right to the qps)

**Comment by He-Pin** [comment]
@franz1981 Which means the average rt seems higher than the pooled bytebuf allocator, but yes, this time no oom.

**Comment by franz1981** [comment]
I have filled #14912 which is a wild guess for a fairly common case which could be the culprit, but if you can collect CPU profiling data to compare adaptive vs jemalloc it would speed the investigation a lot 🙏

**Comment by He-Pin** [comment]
@franz1981 Thanks for the update. I have to switch to another project, but I will try it once I find the time. The payload is mostly the result of  SSR, so the payload is quite large (always the same webpage)
We are using UDS to receive the rendering result and resending it to clients.

**Comment by dpy1123** [comment]
hi team, this issue happened again in our production. Still rss growing can cause application oom killed by system.

We are using netty version 4.1.118.Final, with AdaptiveByteBufAllocator.

<img width="855" alt="Image" src="https://github.com/user-attachments/assets/642dbbdd-cd45-4276-971f-c1c7ceffbcc3" />

![Image](https://github.com/user-attachments/assets/ba3a928c-4d33-4c8d-b8e9-22e704932cf0)

**Comment by yawkat** [comment]
@dpy1123 please note that 4.1.118 does not have leak detection for the adaptive allocator, yet. https://github.com/netty/netty/pull/14949

**Comment by dpy1123** [comment]
> [@dpy1123](https://github.com/dpy1123) please note that 4.1.118 does not have leak detection for the adaptive allocator, yet. [#14949](https://github.com/netty/netty/pull/14949)

yes, we noticed that will be included in 4.1.120, but now still in progress. We can copy the modification and test offline, to see if we can reproduce this issuse.

**Comment by linking12** [comment]
sorry, we need repoen this issue, we found the problem was not solved. Three weeks after the launch, the system still had OOM due to memory not being released.

**Comment by franz1981** [comment]
@linking12 I still believe you should profile for leaks (via async profiler) or via jemalloc leak profiling - but it will miss the Java stack frame.
So only async profiler (version 4 got it ootb) seems the most viable option to me.
Using it can tell us which operation can be problematic and required to keep on increasing memory.
If it is not due to fragmentation (which shouldn't be) maybe is just a bug on enlarging buffers or other common operations

**Comment by laosijikaichele** [comment]
@linking12 Which OS allocator do you use? glibc or jemalloc?

**Comment by franz1981** [comment]
@laosijikaichele I remember he mentioned jemalloc

**Comment by dreamlike-ocean** [comment]
> [@linking12](https://github.com/linking12) What OS allocator do you use? glibc or jemalloc?

we use jemalloc

**Comment by linking12** [comment]
@franz1981  @laosijikaichele  
Yes, we switched to jemalloc before reusing adaptitive to avoid memory fragmentation issues, but it still happened. We also tried to use jemalloc's tools for analysis, but unfortunately, jemalloc's tools can only diagnose instantaneous values, and we can't catch the zero boundary point. The same is true for async profiler. This problem takes 3 weeks to appear, and only occasionally in one or two nodes, so any suggestions?

**Comment by franz1981** [comment]
@dpy1123 just to make sure: your problem is the same as @linking12 ? You both work for the same company and we still talk about the same issue?

Because in https://github.com/netty/netty/issues/14598#issuecomment-2797465990 it doesn't look like a problem which can arise in 2 weeks, but few hours...

**Comment by linking12** [comment]
> [@dpy1123](https://github.com/dpy1123) just to make sure: your problem is the same as [@linking12](https://github.com/linking12) ? You both work for the same company and we still talk about the same issue?
> 
> Because in [#14598 (comment)](https://github.com/netty/netty/issues/14598#issuecomment-2797465990) it doesn't look like a problem which can arise in 2 weeks, but few hours...

@franz1981 

@dpy1123 @dreamlike-ocean @linking12 are same company & same team for a access gateway,  we ara talking same issue;

**Comment by laosijikaichele** [comment]
I noticed there was a [reproduce code](https://github.com/netty/netty/issues/14269#issue-2493316288) which set  `SO_SNDBUF` to 4MiB:

> .childOption(ChannelOption.SO_SNDBUF, 4 * 1024 * 1024)

Do you also explicitly set `SO_SNDBUF` in your production code? 
@linking12

**Comment by dreamlike-ocean** [comment]
> I noticed there was a [reproduce code](https://github.com/netty/netty/issues/14269#issue-2493316288) which set `SO_SNDBUF` to 4MiB:
> 
> > .childOption(ChannelOption.SO_SNDBUF, 4 * 1024 * 1024)
> 
> Do you also explicitly set `SO_SNDBUF` in your production code?
> 
> [@linking12](https://github.com/linking12)

yes,we do.Will SO_SNDBUF cause OOM? I think the configuration of SO_SNDBUF only affects the TCP kernel send buffer size, right? By the way, we’re using Netty with epoll.

**Comment by laosijikaichele** [comment]
The bigger `SO_SNDBUF` set, the underlying chunk will be released more delayed, which may cumulate/increase the creation of new chunk frequency, so will consume more memory.

In addition, `SO_SNDBUF` might better not manually set. see:
https://github.com/apache/rocketmq/issues/3825

@dreamlike-ocean

**Comment by chrisvest** [comment]
@linking12 When this happens, is there any possibility that you could set a breakpoint in a Java debugger, pause the JVM, and run some code? Or is that totally out of the question?

I thinking of making some utilities to provide greater introspection into what the buffers are doing, but I'd also like to avoid adding any overhead when we're not actively investigating issues like this one.

**Comment by linking12** [comment]
We have some new findings, which may not be adaptive issues. If we are sure, we will update the comment here.

**Comment by Peak-Song** [comment]
@linking12 Hi, notice you're using JDK17, have you set `-add-opens=java.base/java.nio=ALL-UNNAMED` & `-io.netty.tryReflectionSetAccessible` in your JVM options？[this article](https://blog.csdn.net/jdcdev_/article/details/132843927) which have meet similar problem may help

**Comment by He-Pin** [comment]
@Peak-Song So the problem you think it's lack of backpressure, then why the pooled bytebuf allocator will not cause the problem?

**Comment by He-Pin** [comment]
@dreamlike-ocean, just report back to me, it's a slow traffic attack

**Comment by dreamlike-ocean** [comment]
Yes, we are taking some measures to fix the currently leak issues in our code and are monitoring the results. We will keep updating you as we have more information.

**Comment by linking12** [comment]
We can now confirm that this issue should not be related to Adaptive. After I reopened this issue, we changed Adaptive to Pooled, but OOM still occurred occasionally. We finally analyzed it and confirmed that it was our own code problem. The main reason for the issue was that we experienced slow h2 attack, which caused the slow attack request to not be released correctly when h2 connection go away;

I am sorry that many projects are afraid to use Adaptive because of this post.  I hope that after seeing this comment, people will have more confidence in Adaptive.
